test/e2e/apps/deployment.go

```diff
@@ -785,6 +788,127 @@ func testDeploymentsControllerRef(f *framework.Framework) {
+	firstRS, err = c.ExtensionsV1beta1().ReplicaSets(ns).Get(firstRS.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	secondRS, err := deploymentutil.GetNewReplicaSet(deployment, c.ExtensionsV1beta1())
+	Expect(err).NotTo(HaveOccurred())
+
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), firstRS)
+	Expect(err).NotTo(HaveOccurred())
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), secondRS)
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that first rollout's replicaset has minimum required number of available replicas")
+	if firstRS.Status.AvailableReplicas < minAvailableReplicas {
+		Expect(fmt.Errorf("observed %d available replicas, less than min required %d", firstRS.Status.AvailableReplicas, minAvailableReplicas)).NotTo(HaveOccurred())
+	}
+
+	availableReplicas := replicas - deployment.Spec.Strategy.RollingUpdate.MaxUnavailable.IntVal
+	framework.Logf("Waiting for first rollout's replicaset to have %d available replicas", availableReplicas)
+	err = framework.WaitForReplicaSetAvailableReplicas(c.ExtensionsV1beta1(), firstRS, availableReplicas)
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that second rollout's replicaset does not have any available replicas")
+	Expect(secondRS.Status.AvailableReplicas).Should(Equal(int32(0)))
+
+	newReplicas := int32(15)
+	framework.Logf("Scaling up the deployment %q from %d to %d", deploymentName, replicas, newReplicas)
+	deployment, err = framework.UpdateDeploymentWithRetries(c, ns, deployment.Name, func(update *extensions.Deployment) {
+		update.Spec.Replicas = &newReplicas
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	framework.Logf("Waiting for observed generation %d", deployment.Generation)
+	Expect(framework.WaitForObservedDeployment(c, ns, deploymentName, deployment.Generation)).NotTo(HaveOccurred())
+
+	framework.Logf("Verifying that the replicasets of deployment %q have desired number of replicas", deploymentName)
+	firstRS, err = c.ExtensionsV1beta1().ReplicaSets(ns).Get(firstRS.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	secondRS, err = c.ExtensionsV1beta1().ReplicaSets(ns).Get(secondRS.Name, metav1.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), firstRS)
+	Expect(err).NotTo(HaveOccurred())
+	err = framework.WaitForReplicaSetDesiredReplicas(c.ExtensionsV1beta1(), secondRS)
+	Expect(err).NotTo(HaveOccurred())
+
+	numReplicaDiff := newReplicas - replicas
+	if numReplicaDiff < 0 {
+		numReplicaDiff = -numReplicaDiff
+	}
+	biggerProportion := (numReplicaDiff + 1) / 2
+	availableReplicas = availableReplicas + biggerProportion
+	framework.Logf("Verifying that first rollout's replicaset has %d available replicas", availableReplicas)
+	err = framework.WaitForReplicaSetAvailableReplicas(c.ExtensionsV1beta1(), firstRS, availableReplicas)
```

Comment:
Proportional scaling scales active RSes, i.e. change their `spec.replicas`, not their available replicas. Why wait for available replicas here?

Reply:
I was thinking that having the check to pass for `.availableReplicas` meant that `.spec.replicas` check would pass too. In fact, it isn't. The best is to first test _both_: first `.spec.replicas` and then `.availableReplicas`.

Reply:
As sync by person, checking for `.availableReplicas` is not necessary because it is covered by rolling update tests. What we care about is `.spec.replicas` has the intended number.